### ISSUE TITLE
vue/require-default-prop complains about optional props when using typescript interface, withDefaults

### ISSUE THREAD

**Issue by partap** [vuejs/eslint-plugin-vue#1914]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 8.17.0
- **eslint-plugin-vue version:** 9.1.1
- **Node version:** 16.13.2
- **Operating System:** MacOS 10.15.7

**Please show your full configuration:**

(Default file generated by Quasar, with standard & vue3-recommended base rules)
```json5
module.exports = {
  // https://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy
  // This option interrupts the configuration hierarchy at this file
  // Remove this if you have an higher level ESLint config file (it usually happens into a monorepos)
  root: true,

  // https://eslint.vuejs.org/user-guide/#how-to-use-a-custom-parser
  // Must use parserOptions instead of "parser" to allow vue-eslint-parser to keep working
  // `parser: 'vue-eslint-parser'` is already included with any 'plugin:vue/**' config and should be omitted
  parserOptions: {
    parser: require.resolve('@typescript-eslint/parser'),
    extraFileExtensions: [ '.vue' ]
  },

  env: {
    browser: true,
    es2021: true,
    node: true,
    'vue/setup-compiler-macros': true
  },

  // Rules order is important, please avoid shuffling them
  extends: [
    // Base ESLint recommended rules
    // 'eslint:recommended',

    // https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#usage
    // ESLint typescript rules
    'plugin:@typescript-eslint/recommended',

    // Uncomment any of the lines below to choose desired strictness,
    // but leave only one uncommented!
    // See https://eslint.vuejs.org/rules/#available-rules
    // 'plugin:vue/vue3-essential', // Priority A: Essential (Error Prevention)
    // 'plugin:vue/vue3-strongly-recommended', // Priority B: Strongly Recommended (Improving Readability)
    'plugin:vue/vue3-recommended', // Priority C: Recommended (Minimizing Arbitrary Choices and Cognitive Overhead)

    'standard'

  ],

  plugins: [
    // required to apply rules which need type information
    '@typescript-eslint',

    // https://eslint.vuejs.org/user-guide/#why-doesn-t-it-work-on-vue-files
    // required to lint *.vue files
    'vue'

  ],

  globals: {
    ga: 'readonly', // Google Analytics
    cordova: 'readonly',
    __statics: 'readonly',
    __QUASAR_SSR__: 'readonly',
    __QUASAR_SSR_SERVER__: 'readonly',
    __QUASAR_SSR_CLIENT__: 'readonly',
    __QUASAR_SSR_PWA__: 'readonly',
    process: 'readonly',
    Capacitor: 'readonly',
    chrome: 'readonly'
  },

  // add your custom rules here
  rules: {

    // allow async-await
    'generator-star-spacing': 'off',
    // allow paren-less arrow functions
    'arrow-parens': 'off',
    'one-var': 'off',
    'no-void': 'off',
    'multiline-ternary': 'off',

    'import/first': 'off',
    'import/namespace': 'error',
    'import/default': 'error',
    'import/export': 'error',
    'import/extensions': 'off',
    'import/no-unresolved': 'off',
    'import/no-extraneous-dependencies': 'off',

    // The core 'import/named' rules
    // does not work with type definitions
    'import/named': 'off',

    'prefer-promise-reject-errors': 'off',

    quotes: ['warn', 'single', { avoidEscape: true }],

    // this rule, if on, would require explicit return type on the `render` function
    '@typescript-eslint/explicit-function-return-type': 'off',

    // in plain CommonJS modules, you can't use `import foo = require('foo')` to pass this rule, so it has to be disabled
    '@typescript-eslint/no-var-requires': 'off',

    // The core 'no-unused-vars' rules (in the eslint:recommended ruleset)
    // does not work with type definitions
    'no-unused-vars': 'off',

    // allow debugger during development only
    'no-debugger': process.env.NODE_ENV === 'production' ? 'error' : 'off'
  }
}

```

**What did you do?**	

```vue
<script setup lang="ts">
interface Props {
  val?: unknown
  flag?: boolean
}
const props = withDefaults(defineProps<Props>(), {
  flag: false
})

</script>
```

**What did you expect to happen?**
Nothing

**What actually happened?**

```
$ eslint --ext .js,.ts,.vue ./

/Users/partap/src/eslint-vue-prop-defaults/src/components/DumpValue.vue
  3:3  warning  Prop 'val' requires default value to be set  vue/require-default-prop

✖ 1 problem (0 errors, 1 warning)
```

**Repository to reproduce this issue** 
https://github.com/partap/eslint-vue-prop-defaults

If I specify `val: undefined` in the object passed to `withDefaults()` the warning goes away, but I _shouldn't_ have to specify that an optional prop defaults to undefined, right?

This only happens when using `withDefaults` ...but if I *don't* use `withDefaults` then the "flag" prop is no longer optional. The "flag" prop is unused here... I only included it as an example that I might have other props that _do_ need a default value.

**Comment by ota-meshi** [comment]
Thank you for posting this issue.

> I shouldn't have to specify that an optional prop defaults to undefined, right?

You don't have to do that. If you don't need this rule, turn it off.

**Comment by cfmj** [comment]
add this rule in '.eslintrc.js' can fix this waring
`
'vue/require-default-prop': ['off']
`

**Comment by mv-go** [comment]
Hey! I'm a bit late to the party, sorry.

I'm wondering, however, why isn't this rule (`'vue/require-default-prop'`) complaining when I'm defining types without the `withDefaults` wrapper?

```ts
export type MyLogoProps = {
  /**
   * Width of the logo in CSS units.
   * Is overridden by `height` prop if both are set.
   */
  width?: string,
  /**
   * Height of the logo in CSS units.
   * Overrides `width` prop if both are set.
   */
  height?: string,
}

// no complaints here
defineProps<MyLogoProps>()

// will complain that `width` is not "defaulted" if declared using `withDefaults`
// withDefaults(
//   defineProps<LionsoulLogoProps>(),
//   {
//     height: '20px',
//   },
// )
```

From [here](https://eslint.vuejs.org/rules/require-default-prop.html)
> This rule requires default value to be set for each props that are not marked as required (except Boolean props).

I'm not a native speaker, but it seems that the rule description could be a touch misleading. To me it could be read as

> All not-required props should have default values, no matter how these props are declared.
